### PR TITLE
Fixing the SerializationDateTimeConversionMode `AlwaysEpochMillisecdsWithSystemDefaultTimezone`

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
# Description

The `CosmosRowConverter` had a bug assuming that ZoneOffsets between two time zones would be constant - in fact in the 19xx years political rules defining summer/winter time have changed multiple times. To get the correct ZoneOffset the right "local" Instant (vs. Instant.now) needs to be used.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by kushagraThapar — APPROVED** [review]
LGTM

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/ChangeFeedOffset.scala [hunk 1]
```diff
@@ -34,7 +34,7 @@ private[cosmos] case class ChangeFeedOffset
   override def json(): String = jsonPersisted
 }
 
-private[spark] object ChangeFeedOffset {
+private[cosmos] object ChangeFeedOffset {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/ChangeFeedOffset.scala:37]: Is this change intentional?
How was it working before?

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/CreateSpark2ContinuationsFromChangeFeedOffset.scala [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+import com.azure.cosmos.implementation.SparkBridgeImplementationInternal.rangeToNormalizedRange
+import com.azure.cosmos.implementation.changefeed.common.ChangeFeedState
+import com.azure.cosmos.implementation.query.CompositeContinuationToken
+import com.azure.cosmos.spark.{ChangeFeedOffset, CosmosClientCache, CosmosClientCacheItem, CosmosClientConfiguration, CosmosConfig, CosmosContainerConfig, Loan}
+import org.apache.spark.sql.api.java.UDF2
+
+import scala.collection.mutable
+
+@SerialVersionUID(1L)
+class CreateSpark2ContinuationsFromChangeFeedOffset extends UDF2[Map[String, String], String, Map[Int, Long]] {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/CreateSpark2ContinuationsFromChangeFeedOffset.scala:16]: Spark2 ?
I didn't get the naming?

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/CreateSpark2ContinuationsFromChangeFeedOffset.scala [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  override def call
+  (
+    userProvidedConfig: Map[String, String],
+    changeFeedOffset: String
+  ): Map[Int, Long] = {
+
+    val effectiveUserConfig = CosmosConfig.getEffectiveConfig(None, None, userProvidedConfig)
+    val cosmosClientConfig = CosmosClientConfiguration(
+      effectiveUserConfig,
+      useEventualConsistency = false)
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/CreateSpark2ContinuationsFromChangeFeedOffset.scala:26]: shouldn't we read this from config as being done at other places?

File: eng/versioning/version_client.txt [hunk 4]
```diff
@@ -88,8 +88,8 @@ com.azure:azure-cosmos;4.33.1;4.34.0-beta.1
 com.azure:azure-cosmos-benchmark;4.0.1-beta.1;4.0.1-beta.1
 com.azure:azure-cosmos-dotnet-benchmark;4.0.1-beta.1;4.0.1-beta.1
 com.azure.cosmos.spark:azure-cosmos-spark_3_2-12;1.0.0-beta.1;1.0.0-beta.1
-com.azure.cosmos.spark:azure-cosmos-spark_3-1_2-12;4.12.1;4.13.0-beta.1
-com.azure.cosmos.spark:azure-cosmos-spark_3-2_2-12;4.12.1;4.13.0-beta.1
+com.azure.cosmos.spark:azure-cosmos-spark_3-1_2-12;4.12.1;4.12.2
```
Comment (kushagraThapar) [inline eng/versioning/version_client.txt:91]: Are we planning to ship this release with azure-cosmos 4.34.0-beta.1 version?

**Review by kushagraThapar — APPROVED** [review]

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/CreateSpark2ContinuationsFromChangeFeedOffset.scala (hunk 2 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/CreateSpark2ContinuationsFromChangeFeedOffset.scala:16]: Spark 2 connnector used a different way of storing changefeed continuations in hdfs. This udf allows ADF to in theory migarte a customer from their Spark 2 to Spark 3 environment and also back in case the customer is hitting some regression.

File: eng/versioning/version_client.txt (hunk 4 above)
Comment (FabianMeiswinkel) [reply eng/versioning/version_client.txt:91]: yes - shaded - beta tag will be renamed to snapshot. And because we shade azure-cosmos we don't also have to release it.

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/CreateSpark2ContinuationsFromChangeFeedOffset.scala (hunk 3 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/udf/CreateSpark2ContinuationsFromChangeFeedOffset.scala:26]: For calls we are making (to metadata) eventual or not is irrelevant.

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/ChangeFeedOffset.scala (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/ChangeFeedOffset.scala:37]: It is intentional - before only classes form the com.azure.cosmos.spark package (or "sub packages") were able to access this internal api. I know extended visibility to all packages under com.azure.cosmos because internal PAI access from one of the Bridge classes is needed.
